core-services/workflow/src/main/java/org/silverpeas/core/workflow/engine/WorkflowEngineImpl.java

```diff
@@ -335,103 +190,80 @@ public void reAssignActors(UpdatableProcessInstance instance, Actor[] unAssigned
       Actor[] assignedActors, User user) throws WorkflowException {
     // Get the process instance
     ProcessInstanceManager instanceManager = WorkflowHub.getProcessInstanceManager();
-    Database db = null;
-    UpdatableHistoryStep step;
-
-    // Get the task manager
-    TaskManager taskManager = WorkflowHub.getTaskManager();
-
-    try {
-      // Get database connection
-      db = WorkflowJDOManager.getDatabase();
-
-      // begin transaction
-      db.begin();
-
-      // Re-load process instance
-      instance =
-          (UpdatableProcessInstance) db.load(ProcessInstanceImpl.class, instance.getInstanceId());
-
-      // first create the history step
-      try {
-        step = (UpdatableHistoryStep) instanceManager.createHistoryStep();
-        step.setUserId(user.getUserId());
-        step.setAction("#reAssign#");
-        step.setActionDate(new Date());
-        step.setUserRoleName("supervisor");
-        step.setResolvedState("DummyStateFromReassignment");
-        step.setActionStatus(0); // To be processed
-
-        // add the new step to the processInstance
-        instance.addHistoryStep(step);
-      } catch (WorkflowException we) {
-        db.rollback();
-        throw new WorkflowException("WorkflowEngineImpl.assignActors", "EX_ERR_REASSIGN_ACTORS",
-            we);
+    String id = instance.getInstanceId();
+    Mutable<UpdatableHistoryStep> step = Mutable.of(null);
+
+    // first create the history step
+    Transaction.performInOne(()-> {
+      UpdatableProcessInstance processInstance = repository.getById(id);
+
+      UpdatableHistoryStep newStep = (UpdatableHistoryStep) instanceManager.createHistoryStep();
+      newStep.setUserId(user.getUserId());
+      newStep.setAction("#reAssign#");
+      newStep.setActionDate(new Date());
+      newStep.setUserRoleName("supervisor");
+      newStep.setResolvedState("DummyStateFromReassignment");
+      // To be processed
+      newStep.setActionStatus(0);
+      newStep.setProcessInstance(processInstance);
+
+      historyStepRepository.save((HistoryStepImpl) newStep);
+
+      step.set(newStep);
+      return newStep;
+    });
+
+    // add the new step to the processInstance
+    instance.addHistoryStep(step.get());
+
+    Transaction.performInOne(()-> {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce this lambda expression number of lines from 27 to at most 20. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1188)